src/SourceTerms/Leakage/CPU_Src_Leakage.cpp

```diff
@@ -0,0 +1,1059 @@
+
+
+
+/********************************************************
+1. Leakage source term
+   --> Enabled by the runtime option "SRC_LEAKAGE"
+
+2. This file is shared by both CPU and GPU
+
+   CUSRC_Src_Leakage.cu -> CPU_Src_Leakage.cpp
+
+3. Four steps are required to implement a source term
+
+   I.   Set auxiliary arrays
+   II.  Implement the source-term function
+   III. [Optional] Add the work to be done every time
+        before calling the major source-term function
+   IV.  Set initialization functions
+
+4. The source-term function must be thread-safe and
+   not use any global variable
+********************************************************/
+
+
+
+// =======================
+// I. Set auxiliary arrays
+// =======================
+
+//-------------------------------------------------------------------------------------------------------
+// Function    :  Src_SetAuxArray_Leakage
+// Description :  Set the auxiliary arrays AuxArray_Flt/Int[]
+//
+// Note        :  1. Invoked by Src_Init_Leakage()
+//                2. AuxArray_Flt/Int[] have the size of SRC_NAUX_LEAKAGE defined in Macro.h (default = 11)
+//                3. Add "#ifndef __CUDACC__" since this routine is only useful on CPU
+//
+// Parameter   :  AuxArray_Flt/Int : Floating-point/Integer arrays to be filled up
+//
+// Return      :  AuxArray_Flt/Int[]
+//-------------------------------------------------------------------------------------------------------
+#ifndef __CUDACC__
+void Src_SetAuxArray_Leakage( double AuxArray_Flt[], int AuxArray_Int[] )
+{
+
+   const int    Mode    = (  (SrcTerms.Leakage_NTheta == 1) << 1  ) + (SrcTerms.Leakage_NPhi == 1);
+   const int    NRay    = SrcTerms.Leakage_NTheta * SrcTerms.Leakage_NPhi;
+   const double Ye_Frac = 0.01;
+
+   AuxArray_Flt[SRC_AUX_PNS_X     ] = amr->BoxCenter[0];
+   AuxArray_Flt[SRC_AUX_PNS_Y     ] = amr->BoxCenter[1];
+   AuxArray_Flt[SRC_AUX_PNS_Z     ] = amr->BoxCenter[2];
+   AuxArray_Flt[SRC_AUX_MAXRADIUS ] = SrcTerms.Leakage_RadiusMax;
+   AuxArray_Flt[SRC_AUX_RADMIN_LOG] = int( SrcTerms.Leakage_RadiusMin_Log / SrcTerms.Leakage_BinSize_Radius )
+                                    * SrcTerms.Leakage_BinSize_Radius;
+   AuxArray_Flt[SRC_AUX_DRAD      ] = SrcTerms.Leakage_BinSize_Radius;
+   AuxArray_Flt[SRC_AUX_DTHETA    ] = M_PI / SrcTerms.Leakage_NTheta;
+   AuxArray_Flt[SRC_AUX_DPHI      ] = 2.0 * M_PI / SrcTerms.Leakage_NPhi;
+#  if ( EOS == EOS_NUCLEAR )
+   AuxArray_Flt[SRC_AUX_YEMIN     ] = ( 1.0 + Ye_Frac ) * h_EoS_Table[NUC_TAB_YE][0];
+   AuxArray_Flt[SRC_AUX_YEMAX     ] = ( 1.0 - Ye_Frac ) * h_EoS_Table[NUC_TAB_YE][g_nye-1];
+#  endif
+   AuxArray_Flt[SRC_AUX_VSQR2CODE ] = 1.0 / SQR( UNIT_V );
+
+   AuxArray_Int[SRC_AUX_MODE      ] = Mode;
+   AuxArray_Int[SRC_AUX_NRAD_LIN  ] = int( SrcTerms.Leakage_RadiusMin_Log / SrcTerms.Leakage_BinSize_Radius );
+   AuxArray_Int[SRC_AUX_STRIDE    ] = NRay * NType_Neutrino;
+   AuxArray_Int[SRC_AUX_RECORD    ] = 0;
```

Comment:
What is the purpose of this and the corresponding codes below?

Reply:
To store and return the leakage information in the built-in fields in `fluid[]`, instead of declaring new fields, for recording the leakage information in `Record__CentralQuant` (the filename may be updated in next PR).